File: energy_macro_system/R/run_all.R
# run_all.R
# Energy–Macro Shock Propagation Simulator — run full pipeline (data + modeling)
# Set working directory to energy_macro_system before sourcing.

# Execution order: 01 -> 02 -> 03 -> 04 -> 05 (06 and 07 are called explicitly or from 05)
message("=== Energy–Macro Shock Propagation Simulator — Pipeline ===\n")

# Ensure we are in energy_macro_system (project root for this app)
if (basename(getwd()) != "energy_macro_system") {
  if (dir.exists("energy_macro_system")) setwd("energy_macro_system")
  else if (dir.exists("../energy_macro_system")) setwd("../energy_macro_system")
}
ROOT <- getwd()
R_DIR <- file.path(ROOT, "R")
DATA_DIR <- file.path(ROOT, "data")

# Phase 1: Data ingestion
message("--- Phase 1: Data ingestion ---")
source(file.path(R_DIR, "01_fetch_fred.R"), local = TRUE)
run_fetch_fred(DATA_DIR)
source(file.path(R_DIR, "02_fetch_eia.R"), local = TRUE)
run_fetch_eia(DATA_DIR)
source(file.path(R_DIR, "03_fetch_worldbank.R"), local = TRUE)
run_fetch_worldbank(DATA_DIR)
source(file.path(R_DIR, "04_data_cleaning.R"), local = TRUE)
run_data_cleaning(DATA_DIR, "system_timeseries.csv")

# Phase 2: Parameter estimation
message("\n--- Phase 2: Parameter estimation ---")
source(file.path(R_DIR, "05_parameter_estimation.R"), local = TRUE)
est <- run_parameter_estimation(DATA_DIR)

# Phase 3 & 4: Monte Carlo and stability
message("\n--- Phase 3: Monte Carlo (example) ---")
source(file.path(R_DIR, "06_monte_carlo_engine.R"), local = TRUE)
A <- est$A
Sigma <- est$Sigma
mc <- simulate_system(A, Sigma, shock_vector = NULL, horizon = 24L, n_sim = 1000L)
message("Recession prob (end): ", round(mc$recession_prob_end, 3))
message("Inflation > 4% prob: ", round(mc$inflation_stress_prob, 3))
message("Demand stress metric: ", round(mc$demand_stress_metric, 3))

message("\n--- Phase 4: Stability metrics ---")
source(file.path(R_DIR, "07_stability_metrics.R"), local = TRUE)
stab <- run_stability_metrics(DATA_DIR)
message(stab$stability_condition)
message("Shock amplification factor: ", round(stab$shock_amplification_factor, 4))

# Optional: LLM interpretation of VAR stability (eigenvalues, spectral radius, shock amplification)
tryCatch({
  source(file.path(R_DIR, "08_llm_layer.R"), local = TRUE)
  llm_out <- llm_interpret_var_stability(stab, use_ollama = TRUE, use_openai = TRUE)
  if (llm_out$success && nzchar(llm_out$narrative)) {
    message("\n--- LLM VAR stability interpretation ---")
    message(llm_out$narrative)
  }
}, error = function(e) message("LLM interpretation skipped: ", conditionMessage(e)))

message("\n=== Pipeline complete. ===")

File: energy_macro_system/run_all.py
#!/usr/bin/env python3
# run_all.py — Run full Energy–Macro Shock Propagation Simulator pipeline (Python)

import os
import subprocess
import sys
from pathlib import Path

# Run from energy_macro_system or project root
ROOT = Path(__file__).resolve().parent
if ROOT.name != "energy_macro_system" and (ROOT / "energy_macro_system").exists():
    ROOT = ROOT / "energy_macro_system"
os.chdir(ROOT)
sys.path.insert(0, str(ROOT / "python"))

DATA_DIR = ROOT / "data"
DATA_DIR.mkdir(parents=True, exist_ok=True)


def main():
    print("=== Energy–Macro Shock Propagation Simulator — Pipeline (Python) ===\n")

    # Phase 1: Data ingestion
    print("--- Phase 1: Data ingestion ---")
    import load_env
    load_env.load_dotenv(load_env.find_env_path())

    import fetch_fred
    fetch_fred.run_fetch_fred(str(DATA_DIR))

    import fetch_eia
    fetch_eia.run_fetch_eia(str(DATA_DIR))

    import fetch_worldbank
    fetch_worldbank.run_fetch_worldbank(str(DATA_DIR))

    import data_cleaning
    data_cleaning.run_data_cleaning(str(DATA_DIR))

    # Phase 2: Parameter estimation
    print("\n--- Phase 2: Parameter estimation ---")
    import parameter_estimation
    est = parameter_estimation.run_parameter_estimation(str(DATA_DIR))

    # Phase 3: Monte Carlo
    print("\n--- Phase 3: Monte Carlo (example) ---")
    import monte_carlo_engine
    mc = monte_carlo_engine.simulate_system(est["A"], est["Sigma"], horizon=24, n_sim=1000)
    print(f"Recession prob (end): {mc['recession_prob_end']:.3f}")
    print(f"Inflation > 4% prob: {mc['inflation_stress_prob']:.3f}")
    print(f"Demand stress metric: {mc['demand_stress_metric']:.3f}")

    # Phase 4: Stability
    print("\n--- Phase 4: Stability metrics ---")
    import stability_metrics
    stab = stability_metrics.run_stability_metrics(str(DATA_DIR), est.get("columns"))
    print(stab["stability_condition"])
    print(f"Shock amplification factor: {stab['shock_amplification_factor']:.4f}")

    print("\n=== Pipeline complete. ===")
    print("\n--- Starting Shiny app ---")
    print("Dashboard will open at http://localhost:3838 (press Ctrl+C to stop).\n")
    subprocess.run(
        ["R", "-e", "shiny::runApp('.', port = 3838, launch.browser = TRUE)"],
        cwd=str(ROOT),
    )


if __name__ == "__main__":
    main()
